Switched to generic endian swap

gnawme opened the pull request:
Thanks a lot for making a contribution to SFML! 🙂

Before you create the pull request, we ask you to check the follow boxes. (For small changes not everything needs to ticked, but the more the better!)

* [x] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [x] Have you provided some example/test code for your changes?
* [ ] If you have additional steps which need to be performed list them as tasks!

----

## Description
Switched to use [generic (templated) endian swap method](https://mklimenko.github.io/english/2018/08/22/robust-endian-swap/) that I've used successfully in the past.

This PR is related to the issue #2290

## Tasks

* [x] Tested on Linux
* [ ] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?
Extended the `Packet` stream tests to include `float` and `double`

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,13 +31,37 @@
+///
+////////////////////////////////////////////////////////////
+template <typename T>
+T swapEndian(T& val, typename std::enable_if<std::is_arithmetic<T>::value, std::nullptr_t>::type = nullptr)
```
Why a mutable l-value reference? Why not just copy it? We're only talking about types up to 8 bytes so it's not like that copy is any worse than taking a reference.

ChrisThrasher replied:
This ought to be moved to the .cpp file so that it doesn't become a new part of the API.

ChrisThrasher replied:
I don't think we need SFINAE here and can instead just use a `static_assert`. SFINAE is great when dealing with overload sets but we don't have an overload set here. We just need some way to break compilation if an unacceptable type is used.

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,13 +31,37 @@
+template <typename T>
+T swapEndian(T& val, typename std::enable_if<std::is_arithmetic<T>::value, std::nullptr_t>::type = nullptr)
+{
+    union U
```
I believe we can avoid giving this union a name

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,13 +31,37 @@
+    } src, dst;
+
+    src.val = val;
+    std::reverse_copy(src.raw.begin(), src.raw.end(), dst.raw.begin());
```
Doesn't this constitute type punning through a union and thus is UB?

gnawme replied:
Switched to not use a union.

ChrisThrasher commented on src/SFML/Network/Packet.cpp:
```diff
@@ -200,7 +200,7 @@ Packet& Packet::operator>>(std::uint32_t& data)
     {
         std::memcpy(&data, &m_data[m_readPos], sizeof(data));
-        data = ntohl(data);
+        data = sf::swapEndian(data);
```
```suggestion
        data = swapEndian(data);
```
You can drop the `sf::` namespace whenever you're calling this.

ChrisThrasher commented on test/Network/Packet.test.cpp:
```diff
@@ -97,5 +97,21 @@ TEST_CASE("[Network] sf::Packet")
+
+        SUBCASE("float")
+        {
+            CHECK_PACKET_STREAM_OPERATORS(float(0.f));
+            CHECK_PACKET_STREAM_OPERATORS(float(1.f));
```
```suggestion
            CHECK_PACKET_STREAM_OPERATORS(0.f);
            CHECK_PACKET_STREAM_OPERATORS(1.f);
```

ChrisThrasher commented on test/Network/Packet.test.cpp:
```diff
@@ -97,5 +97,21 @@ TEST_CASE("[Network] sf::Packet")
+
+        SUBCASE("double")
+        {
+            CHECK_PACKET_STREAM_OPERATORS(double(0.0));
+            CHECK_PACKET_STREAM_OPERATORS(double(1.0));
```
```suggestion
            CHECK_PACKET_STREAM_OPERATORS(0.0);
            CHECK_PACKET_STREAM_OPERATORS(1.0);
```

ChrisThrasher commented:
If I understand correct, this PR would make sure bytes always get swapped around but what about the case where the machine's endianness matches network endianness? In that case we don't need to swap bytes around but this would seemingly still swap them.

ChrisThrasher commented:
https://github.com/ChrisThrasher/SFML/commit/26ae62e694ef7866a1fe69d1a0ef91f902b973ae

This inspired me to write tests that make sure we're using the correct network ordering. Let me know if you can run your code with these tests and see if anything fails.

JonnyPtn commented:
I'm not convinced SFML should be doing any reordering - in the vast majority of cases it's going to be little endian machines on both sides, using SFML sockets. In which cases the conversion is just redundant extra work for both sides

I think the only use case this benefits would be people communicating with other services they have no control over, in which case they can still do the conversion themselves?

At best this should be optional, and off by default

gnawme commented:
> [ChrisThrasher@26ae62e](https://github.com/ChrisThrasher/SFML/commit/26ae62e694ef7866a1fe69d1a0ef91f902b973ae)
> 
> This inspired me to write tests that make sure we're using the correct network ordering. Let me know if you can run your code with these tests and see if anything fails.

Where can I access these tests?

ChrisThrasher commented:
https://github.com/SFML/SFML/compare/master...ChrisThrasher:SFML:test_network_order

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,12 +31,55 @@
+////////////////////////////////////////////////////////////
+inline bool isBigEndian()
+{
+    int one{1};
```
```suggestion
    int one = 1;
```
We don't use braces for initialization in this context.

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,12 +31,55 @@
+
+    std::array<std::uint8_t, sizeof(T)> raw;
+
+    auto* cast{reinterpret_cast<std::uint8_t*>(&val)};
```
```suggestion
    auto* cast = reinterpret_cast<std::uint8_t*>(&val);
```

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,12 +31,55 @@
+    std::reverse_copy(raw.begin(), raw.end(), dst.begin());
+    std::memcpy(cast, dst.data(), sizeof(T));
+
+    auto* valp{reinterpret_cast<T*>(cast)};
```
```suggestion
    auto* valp = reinterpret_cast<T*>(cast);
```

ChrisThrasher commented on src/SFML/Network/IpAddress.cpp:
```diff
@@ -62,8 +62,8 @@ std::optional<IpAddress> IpAddress::resolve(std::string_view address)
     // Try to convert the address as a byte representation ("xxx.xxx.xxx.xxx")
-    if (const std::uint32_t ip = inet_addr(address.data()); ip != INADDR_NONE)
-        return IpAddress(ntohl(ip));
+    if (auto ip = inet_addr(address.data()); ip != INADDR_NONE)
```
```suggestion
    if (const auto ip = inet_addr(address.data()); ip != INADDR_NONE)
```

ChrisThrasher replied:
Now that `swapEndian` no longer takes a mutable reference, we can put `const` back where it was removed.

ChrisThrasher commented on src/SFML/Network/IpAddress.cpp:
```diff
@@ -75,10 +75,10 @@ std::optional<IpAddress> IpAddress::resolve(std::string_view address)
         std::memcpy(&sin, result->ai_addr, sizeof(*result->ai_addr));
 
-        const std::uint32_t ip = sin.sin_addr.s_addr;
+        auto ip = sin.sin_addr.s_addr;
```
```suggestion
        const auto ip = sin.sin_addr.s_addr;
```

ChrisThrasher commented on src/SFML/Network/IpAddress.cpp:
```diff
@@ -111,7 +111,8 @@ std::string IpAddress::toString() const
 std::uint32_t IpAddress::toInteger() const
 {
-    return ntohl(m_address);
+    auto address{m_address};
+    return swapEndian(address);
```
```suggestion
    return swapEndian(m_address);
```

ChrisThrasher commented on src/SFML/Network/IpAddress.cpp:
```diff
@@ -128,7 +129,8 @@ std::optional<IpAddress> IpAddress::getLocalAddress()
 
     // Connect the socket to localhost on any port
-    sockaddr_in address = priv::SocketImpl::createAddress(ntohl(INADDR_LOOPBACK), 9);
+    std::uint32_t kcabpool{INADDR_LOOPBACK};
```
```suggestion
    std::uint32_t kcabpool = INADDR_LOOPBACK;
```

ChrisThrasher commented on include/SFML/Network/Socket.hpp:
```diff
@@ -31,12 +31,55 @@
+///
+////////////////////////////////////////////////////////////
+template <typename T>
+T swapEndian(T val)
```
```suggestion
[[nodiscard]] T swapEndian(T val)
```

ChrisThrasher reviewed (changes requested):
A lot of these functions are collapsing down to nearly identical code. I wonder if this new pattern means we can replace a lot of these function overloads with a template and cut out hundreds of lines of code.

ChrisThrasher commented on src/SFML/Network/TcpSocket.cpp:
```diff
@@ -68,7 +68,7 @@ unsigned short TcpSocket::getLocalPort() const
         if (getsockname(getHandle(), reinterpret_cast<sockaddr*>(&address), &size) != -1)
         {
-            return ntohs(address.sin_port);
+            return swapEndian(address.sin_port);
```
I don't think we necessarily need to stop using `ntohs` and related functions. The motivation for this PR is to fix issues with how `sf::Packet` handles floats so I'd rather not touch other classes.

gnawme replied:
Reverted all except `Packet.cpp`

gnawme commented:
> A lot of these functions are collapsing down to nearly identical code. I wonder if this new pattern means we can replace a lot of these function overloads with a template and cut out hundreds of lines of code.

I was thinking about that. Is there a consistent pattern that can be applied to character strings? Those are also not converted to network byte order.

ChrisThrasher commented on src/SFML/Network/Packet.cpp:
```diff
@@ -29,12 +29,57 @@
+
+    std::array<std::uint8_t, sizeof(T)> dst;
+    std::reverse_copy(raw.begin(), raw.end(), dst.begin());
+    std::memcpy(cast, dst.data(), sizeof(T));
+
+    auto* valp = reinterpret_cast<T*>(cast);
+
+    return *valp;
```
If we're copying into the address of `val`, can't we then just return `val` whose bytes have just been rewritten? As written this appears to be starting the lifetime of a type with `reinterpret_cast` which is UB.

ChrisThrasher commented on src/SFML/Network/Packet.cpp:
```diff
@@ -29,12 +29,57 @@
+///
+////////////////////////////////////////////////////////////
+template <typename T>
+T swapEndian(T val)
```
```suggestion
T swapEndian(T value)
```
There's no need to abbreviate words that are already 5 characters. We allow for 120 character columns so vertical space is not a precious resource. Same goes for other abbreviated variable names. Please be descriptive.

ChrisThrasher commented:
> Is there a consistent pattern that can be applied to character strings? Those are also not converted to network byte order.

Each character is one byte which means the endianness doesn't really matter. If each character is >1 byte then I'd expect each character to get its bytes rearranged but not to reverse the string itself.

ChrisThrasher commented:
I'm going to pause on the PR feedback until some other maintainers can weigh in on whether they like reimplementing `sf::Packet`. I believe that we can use a template with a closed set of specializations to simplify how many of these `<<` and `>>` overloads are implemented much like how we explicitly specialize the `sf::Vector2<T>` template for floating point types to avoid having to put quite so much of that code in the header.

ChrisThrasher commented:
SFMLBuildMaster: Build this please.

gnawme commented:
> believe that we can use a template with a closed set of specializations

I imagine that would be the subject of a separate PR

ChrisThrasher commented:
> I imagine that would be the subject of a separate PR

I see it as part of justifying this PR. It's easier to get this approved and merged this if it also means getting to delete huge swaths of code.

gnawme commented:
> I see it as part of justifying this PR.

Any sense of when and how the maintainers want to move forward?

ChrisThrasher commented:
> Any sense of when and how the maintainers want to move forward?

Sorry I'm not sure.

gnawme committed:
Switched to generic endian swap

Added missing <array> header

Addressed review comments

Restricted templated swaps to Packet

Made variable names more descriptive per review

Initial implementation of templated operators

 Changed to if constexpr to work around C4127 error

Conformed initialization style

Removed unused variable

ZXShady commented:
this is UB you cannot just access 2 union members at a time

```cpp
template <typename T>
void SwapEndian(T &val) {
    union U {
        T val;
        std::array<std::uint8_t, sizeof(T)> raw; // UB!
    } src, dst;

    src.val = val; // UB
    std::reverse_copy(src.raw.begin(), src.raw.end(), dst.raw.begin());
    val = dst.val; // UB
}
```

it is UB to access 2 union members in C++ by the standard it is allowed in C (since C99 and explicitly allowed in C11) but not C++ at all but all compilers make this behavoir defined but I would not like SFML to rely on possible UB even if it is supported
